src/v/archival/ntp_archiver_service.cc

```diff
@@ -453,21 +479,46 @@ ntp_archiver::upload_tx(upload_candidate candidate) {
     co_return co_await _remote.upload_manifest(_bucket, manifest, fib);
 }
 
-ss::future<ntp_archiver::scheduled_upload> ntp_archiver::schedule_single_upload(
-  model::offset start_upload_offset, model::offset last_stable_offset) {
+static ss::future<std::vector<ss::rwlock::holder>> get_read_locks(
+  const upload_candidate& candidate, ss::lowres_clock::duration timeout) {
+    auto deadline = std::chrono::steady_clock::now() + timeout;
+    std::vector<ss::future<ss::rwlock::holder>> locks;
+    locks.reserve(candidate.sources.size());
+    std::transform(
+      candidate.sources.begin(),
+      candidate.sources.end(),
+      std::back_inserter(locks),
+      [&deadline](auto& seg) { return seg->read_lock(deadline); });
+    co_return co_await ss::when_all_succeed(locks.begin(), locks.end());
+}
+
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
+    auto start_upload_offset = upload_ctx.start_offset;
+    auto last_stable_offset = upload_ctx.last_offset;
     std::optional<storage::log> log = _partition_manager.log(_ntp);
     if (!log) {
         vlog(_rtclog.warn, "couldn't find log in log manager");
         co_return scheduled_upload{.stop = ss::stop_iteration::yes};
     }
 
-    auto upload = co_await _policy.get_next_candidate(
-      start_upload_offset,
-      last_stable_offset,
-      *log,
-      *_partition->get_offset_translator_state());
+    upload_candidate upload;
+    switch (upload_ctx.upload_kind) {
+    case segment_upload_kind::non_compacted:
+        upload = co_await _policy.get_next_candidate(
+          start_upload_offset,
+          last_stable_offset,
+          *log,
+          *_partition->get_offset_translator_state());
+        break;
+    case segment_upload_kind::compacted:
+        auto m = manifest();
+        upload = co_await _policy.get_next_compacted_segment(
```

Comment:
I might have missed this in the PR, but when I looked at the code when writing the code, the offset translator was only used within archival policy [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/archival/archival_policy.cc#L228) to handle the case where the segment was open, to avoid upload where kafka offset had not increased. 

Since this is not the case for compacted segments because they are always closed, I did not use the translator. 

The delta offset is set [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/archival/ntp_archiver_service.cc#L577) when we create the scheduled upload object at the end of `ntp_archiver::schedule_single_upload`.  This part is unchanged as it only relies on base offset.

I can only see `delta_offset_end` in serialization and deserialization. Is it something that needs to be changed for this case or added somewhere?

Reply:
OK, got it